tests/akash/matrix.futil

```diff
@@ -0,0 +1,160 @@
+import "primitives/core.futil";
+import "primitives/binary_operators.futil";
+import "primitives/memories/comb.futil";
+
+// 4x4 matrix multiply
+component main(@go go: 1) -> (@done done: 1) {
+  cells {
+    @external A = comb_mem_d2(32, 4, 4, 3, 3);
+    @external B = comb_mem_d2(32, 4, 4, 3, 3);
+    @external C = comb_mem_d2(32, 4, 4, 3, 3);
+
+    mult = std_mult_pipe(32);
+    add = std_add(32);
+    smalladd = std_add(3);
+    
+    acc = std_reg(32);
+
+    // matrix entries
+    a = std_reg(32);
+    b = std_reg(32);
+
+    // iterators
+    i = std_reg(3);
+    j = std_reg(3);
+    k = std_reg(3);
+
+    // lt's
+    lt_i = std_lt(3);
+    lt_j = std_lt(3);
+    lt_k = std_lt(3);
+  }
+
+  wires {
+
+    group zero_acc {
+      acc.in = 32'd0;
+      acc.write_en = 1'b1;
+      zero_acc[done] = acc.done;
+    }
+
+    group zero_i {
+      i.in = 3'd0;
+      i.write_en = 1'b1;
+      zero_i[done] = i.done;
+    }
+
+    group zero_j {
+      j.in = 3'd0;
+      j.write_en = 1'b1;
+      zero_j[done] = j.done;
+    }
+
+    group zero_k {
+      k.in = 3'd0;
+      k.write_en = 1'b1;
+      zero_k[done] = k.done;
+    }
+
+    comb group cond_i {
+      lt_i.left = i.out;
+      lt_i.right = 3'd4;
+    }
+
+    comb group cond_j {
+      lt_j.left = j.out;
+      lt_j.right = 3'd4;
+    }
+
+    comb group cond_k {
+      lt_k.left = k.out;
+      lt_k.right = 3'd4;
+    }
+
+    group incr_i {
+      smalladd.left = i.out;
+      smalladd.right = 3'd1;
+      i.in = smalladd.out;
+      i.write_en = 1'b1;
+      incr_i[done] = i.done;
+    }
+
+    group incr_j {
+      smalladd.left = j.out;
+      smalladd.right = 3'd1;
+      j.in = smalladd.out;
+      j.write_en = 1'b1;
+      incr_j[done] = j.done;
+    }
+
+    group incr_k {
+      smalladd.left = k.out;
+      smalladd.right = 3'd1;
+      k.in = smalladd.out;
+      k.write_en = 1'b1;
+      incr_k[done] = k.done;
+    }
+
+    group read_A {
+      A.addr0 = i.out;
+      A.addr1 = k.out;
+      a.write_en = 1'b1;
```

Comment:
I actually don't think you need to set `write_en` to high when reading from a combinational memory. Try just dropping this line? I bet `read_data` will just have `A[i][j]` waiting for you. When you move to `seq_mem`, you'll need to set a different flag, `content_en`, to high.

Reply:
I'm reading from memory `A` and writing that value into register `a` (apologies, the names are not great). So, `write_en` here is for the register, which I believe I'd need since I'm writing to it.

Thanks for the heads up about `content_en`!

Reply:
Ah silly me, right you are. Resolving this!